src/graph/plugin_host.rs

```diff
@@ -1168,64 +1158,32 @@ impl Drop for PluginInstanceHostAudioThread {
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
 #[repr(u32)]
 pub(crate) enum PluginState {
-    /// The plugin is inactive, only the main thread uses it
+    // TODO: this state shouldn't be able to exist for the Audio thread
+    /// The plugin is inactive, only the main thread uses it.
     Inactive = 0,
 
-    /// Activation failed
+    /// Activation failed.
     InactiveWithError = 1,
 
-    /// The plugin is active and sleeping, the audio engine can call start_processing()
-    ActiveAndSleeping = 2,
-
-    /// The plugin is processing
-    ActiveAndProcessing = 3,
-
-    /// The plugin is processing, but will be put to sleep the next time all input buffers
-    /// are silent.
-    ActiveAndWaitingForQuiet = 4,
-
-    /// The plugin did process but is in error
-    ActiveWithError = 5,
+    /// The plugin is active. It may or may not be processing right now.
+    Active = 2,
 
-    /// The plugin audio thread is waiting to be dropped.
-    WaitingToDrop = 6,
+    /// The main thread is waiting for the audio thread to drop the plugin's audio processor.
+    WaitingToDrop = 3,
 
-    /// The plugin is not used anymore by the audio engine and can be deactivated on the main
+    /// The plugin is not used anymore by the audio engine and can be deactivated on the main.
     /// thread
-    DroppedAndReadyToDeactivate = 7,
-}
-
-impl PluginState {
-    pub fn is_active(&self) -> bool {
-        !matches!(
-            self,
-            PluginState::Inactive
-                | PluginState::InactiveWithError
-                | PluginState::WaitingToDrop
-                | PluginState::DroppedAndReadyToDeactivate
-        )
-    }
-
-    pub fn is_processing(&self) -> bool {
-        matches!(self, PluginState::ActiveAndProcessing | PluginState::ActiveAndWaitingForQuiet)
-    }
-
-    pub fn is_sleeping(&self) -> bool {
-        *self == PluginState::ActiveAndSleeping
-    }
+    DroppedAndReadyToDeactivate = 4,
 }
 
 impl From<u32> for PluginState {
     fn from(s: u32) -> Self {
         match s {
             0 => PluginState::Inactive,
             1 => PluginState::InactiveWithError,
-            2 => PluginState::ActiveAndSleeping,
-            3 => PluginState::ActiveAndProcessing,
-            4 => PluginState::ActiveAndWaitingForQuiet,
-            5 => PluginState::ActiveWithError,
```

Comment:
Why did these get removed? The first three handle the logic of automatically putting a plugin to sleep.

Although I do agree that the `PluginState::ActiveWithError` case is not really necessary.

Reply:
These have been moved to a new state enum `ProcessingState` here: [src/graph/plugin_host.rs:731](https://github.com/MeadowlarkDAW/dropseed/pull/11/files#diff-9c12b615947d3899cccfd29847105f15889ce0f21bf1d73006523261eebf36f6R731)

I did this because I realized the main thread doesn't care about any of these, it just needs to know if the plugin is active or not when it sends a deactivation request. Only the audio thread has to manage anything related to the processing lifetime.

That change makes the state machine much for straightforward now that these are a hard subset of the active state, and as a bonus it saves on a couple unsynchronized atomic operations, since now the audio thread's processing state cannot interfere with requests from the main thread anymore. :slightly_smiling_face: